tests/cpgrid/LgrChecks.hpp

```diff
@@ -0,0 +1,466 @@
+                BOOST_CHECK_EQUAL(localId, globalId);
+            }
+
+            levelOrLeafIds_set.insert(localId);
+
+            if (!isLeaf) {
+                if (auto it = leafElementLocalId.find(localId); it != leafElementLocalId.end()) {
+                    BOOST_CHECK(it->second.getLevelElem() == element);
+                }
+                const int idx = indexSet.index(element);
+                BOOST_CHECK_EQUAL(idx, element.index());
+            }
+
+            BOOST_CHECK_EQUAL(grid.globalIdSet().id(element),
+                              data[element.level()]->globalIdSet().id(element.getLevelElem()));
+        }
+
+        BOOST_CHECK_EQUAL(static_cast<int>(levelOrLeafIds_set.size()), data[level]->size(0));
+
+        for (const auto& vertex : vertices) {
+            const int localId = localIdSet.id(vertex);
+            const int globalId = globalIdSet.id(vertex);
+
+            if (isSerialRun) {
+                BOOST_CHECK_EQUAL(localId, globalId);
+            }
+
+            levelOrLeafIds_set.insert(localId);
+
+            if (!isLeaf) {
+                if (auto it = leafVertexLocalId.find(localId); it != leafVertexLocalId.end()) {
+                    BOOST_CHECK(it->second.geometry().center() == vertex.geometry().center());
+                }
+            }
+        }
+
+        BOOST_CHECK_EQUAL(static_cast<int>(levelOrLeafIds_set.size()), data[level]->size(0) + data[level]->size(3));
+    }
+}
+
+void Opm::checkLgrNameToLevel(const Dune::CpGrid& grid,
+                              const std::vector<std::string>& lgr_name_vec)
+{
+    BOOST_CHECK_EQUAL( grid.maxLevel(), static_cast<int>(lgr_name_vec.size())); // If there is no nested refinement
+
+    const auto& lgrNameToLevel =  grid.getLgrNameToLevel();
+    for (const auto& [lgrName, level] : lgrNameToLevel) {
+        if (level){
+            BOOST_CHECK_EQUAL( lgrName, lgr_name_vec[level-1]);
+        }
+        else {
+            BOOST_CHECK_EQUAL( lgrName, "GLOBAL");
+        }
+    }
+}
+
+void Opm::checkVertexAndFaceIndexAreNonNegative(const Dune::CpGrid& grid)
+{
+    const auto& leafGridView = grid.leafGridView();
+    for (const auto& element : elements(leafGridView)) {
+        BOOST_CHECK_EQUAL( element.subEntities(3), 8);
+        for (int i = 0; i < 8; ++i){
+            BOOST_CHECK( element.subEntity<3>(i).index() >= 0); // valid vertex index
+        }
+        for (const auto& intersection : Dune::intersections(grid.leafGridView(), element)) {
+            BOOST_CHECK( intersection.id() >= 0); // valid face index
+        }
+    }
+}
+
+void Opm::checkFaceHas4VerticesAndMax2NeighboringCells(const Dune::CpGrid& grid,
+                                                       const std::vector<std::shared_ptr<Dune::cpgrid::CpGridData>>& data)
+{
+    /** Add level grid check */
+
+    const int numFaces = grid.numFaces();
+    for (int face = 0; face < numFaces; ++face){
+        BOOST_CHECK_EQUAL( grid.numFaceVertices(face), 4);
```

Comment:
Just a note: For these tests which are on cartesian grids this holds. In general the can have more veritices as a "face" (actually theses are intersection of two cells) does not need to cover the whole side of an element.